src/bin/identities/CommandPermissions.ts

```diff
@@ -56,7 +56,7 @@ class CommandPermissions extends CommandPolykey {
           logger: this.logger.getChild(WebSocketClient.name),
         });
         pkClient = await PolykeyClient.createPolykeyClient({
-          streamFactory: () => webSocketClient.startConnection(),
+          streamFactory: (ctx) => webSocketClient.startConnection(ctx),
```

Comment:
We call whatever is creating a new stream, so yes, for the quic system we'd call the `streamNew` method. The life-cycle of the connection itself would be maintained separately here.

Reply:
The problem here is:
1. We want to be able to `readable.cancel` or `writable.abort` due to timer deadline which means the stream is destroyed due to idleness.
2. You cannot just use `transformStream.readable.cancel()` without ownership, if called it results in a locked error.
3. The `pipeTo` and `pipeThrough` does support the abort signal, however attempting to use this does not propagate the original reason across the composed streams, instead it would use an abortion error.
Therefore the solution is to have a way to abort the original stream directly.
2 options:
1. Put an `cancel` method into the a derived `ReadableWritablePair` type. We can call this `RPCStream` interface.
2. Demand that the `streamFactory` returns a `PromiseCancellable<[ReadableWritablePair, cancel]>`.
Option 1 is actually the same way we did this with the `Task` type in `src/tasks` domain.
Let's go with option 1.
Create a `RPCStream` interface type with the `cancel` method.
Update all the RPC usages of `ReadableWritablePair` to instead use `RPCStream`.
Then update the transport layer stream factories to produce `RPCStream`.
That also means that `streamFactory(ctx)` now can preserve the original meaning of `timedCancellable`.
That means you can use the `timedCancellable` HOF wrapper around `streamFactory`.
Furthermore the transport layer has to understand `ctx.signal` is used for the lifetime of the asynchronous creation, not the lifetime of the stream itself.

Reply:
Also because the context decorators are not a separate package. And I think there was a reason why this was the case, we cannot use the decorators, instead the HOF has to be used.

But `timedCancellable` HOF should give you back `PromiseCancellable`.

Meaning the `StreamFactory: (ctx: ContextTimed) => Promise<RPCStream<...>>`, because the wrapping happens at the RPC domain.

Note that using the `timedCancellable` allows you to configure the defaults during the wrapping.

Reply:
This change should work. Only concern now is that we have two timeouts, One handled by the TimedCancelable logic for creating the stream, and a separate one for the refresh.

I'm not sure how to get the keep alive timeout. Do I provide a separate option for a keep alive timeout timer separate from the CTX?

Reply:
The first timeout is for acquiring the stream.

The second timeout is to keep the stream alive.

Note that when a RPC call is made, you have a timeout for the entire operation. You pass the same timer down to acquisition of the stream... then the second timeout for the entire keepalive. So from the client's perspective, these are just 2 operations that could happen.

On the server side, you only have the second timeout, the first timeout does not ever exist.

Reply:
As discussed. The timers will be shared for keep alive and stream timeout. So same as currently designed.

Reply:
This has been resolved.